### PR TITLE
Creating S3 bucket on condition

### PR DESCRIPTION
**Pull request by ArvinthC3000**
Signed-off-by: Arvinth C <arvinth.chandrasekaran@progress.com>
Signed-off-by: Tejaswi B <bondila.venkatatejaswi@progress.com>

### :nut_and_bolt: Description: What code changed, and why?
For Automate HA as a product, when the user gives an existing bucket name in the config, it should be added to the configuration. 

Other changes:
- Adding names to Volumes
- Making delete_on_termination configurable
- Updating cleanup scenarios

Note:
1. On Provisioning, terraform will not add a random string at the end of the bucket name
2. S3 Bucket name is mandatory

### :chains: Related Resources
Jira Ticket: https://chefio.atlassian.net/browse/MADROX-427

### :athletic_shoe: How to Build and Test the Change

Use the following deployment package for testing:
`arvinth/automate-ha-deployment/0.1.0/20230120181343`

### :white_check_mark: Checklist

**All PRs** must tick these:

- [ ] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [ ] All commits signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

With occasional exceptions, all PRs **from Progress employees** must tick these:

- [ ] Is the code clear? *(complicated code or lots of comments--subdivide and use well-named methods, meaningful variable names, etc.)*
- [ ] Consistency checked? *(user notifications, user prompts, visual patterns, code patterns, variable names)*
- [ ] Repeated code blocks eliminated? *(adapt and reuse existing components, blocks, functions, etc.)*
- [ ] Spelling, grammar, typos checked? *(at a minimum use `make spell` in any component directory)*
- [ ] Code well-formatted? *(indents, line breaks, etc. improve rather than hinder readability)*

All PRs **from Progress employees** should tick these if appropriate:

- [ ] Tests added/updated? (all new code needs new tests)
- [ ] Docs added/updated? (all customer-facing changes)

*Please add a note next to any checkbox above if you are NOT ticking it.*

### CODE REVIEW

File: components/automate-cli/cmd/chef-automate/automateHAAwsDeployment.go [hunk 1]
```diff
@@ -146,6 +145,9 @@ func (a *awsDeployment) validateConfigFields() *list.List {
 	if len(a.config.Architecture.ConfigInitials.BackupMount) < 1 {
 		errorList.PushBack("Invalid or empty backup_mount")
 	}
+	if a.config.Architecture.ConfigInitials.BackupConfig == "s3" && len(a.config.Architecture.ConfigInitials.S3BucketName) < 1 {
```
Comment (ArvinthC3000) [inline components/automate-cli/cmd/chef-automate/automateHAAwsDeployment.go:148]: add strip logic

File: terraform/a2ha-terraform/modules/s3/templates/createS3Bucket.sh [hunk 2]
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env bash
+
+echo $1
+echo $2
```
Comment (punitmundra) [inline terraform/a2ha-terraform/modules/s3/templates/createS3Bucket.sh:4]: what is $2 ?

File: terraform/a2ha-terraform/modules/s3/inputs.tf [hunk 3]
```diff
@@ -1,3 +1,8 @@
+variable "aws_region" {
+  default     = ""
+  description = "The name of the selected AWS region / datacenter."
+}
+
```
Comment (punitmundra) [inline terraform/a2ha-terraform/modules/s3/inputs.tf:5]: is the terrafrom is not using aws_s3_bucketName then  we can remove it

Comment (punitmundra) [inline terraform/a2ha-terraform/modules/s3/inputs.tf:5]: is the terrafrom is not using aws_s3_bucketName then  we can remove it

File: terraform/a2ha-terraform/modules/s3/templates/createS3Bucket.sh (hunk 2 above)
Comment (ArvinthC3000) [reply terraform/a2ha-terraform/modules/s3/templates/createS3Bucket.sh:4]: $2 is a aws_region variable used for the flag`--create-bucket-configuration LocationConstraint=$2`

File: terraform/a2ha-terraform/modules/s3/inputs.tf (hunk 3 above)
Comment (ArvinthC3000) [reply terraform/a2ha-terraform/modules/s3/inputs.tf:5]: aws_s3_bucketName is used in terraform 'null_resource.createS3bucket' resource

Comment (ArvinthC3000) [reply terraform/a2ha-terraform/modules/s3/inputs.tf:5]: aws_s3_bucketName is used in terraform 'null_resource.createS3bucket' resource

File: terraform/a2ha-terraform/modules/s3/templates/createS3Bucket.sh [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+if hab pkg exec core/aws-cli aws s3 ls s3://$1; then
+  echo "Bucket already exists"
+else
+  export BUCKET_CREATION=$(hab pkg exec core/aws-cli aws s3api create-bucket --bucket $1  --acl private --create-bucket-configuration LocationConstraint=$2)
+  echo "$BUCKET_CREATION"
+  if echo "$BUCKET_CREATION" | grep -q "http://$1.s3.amazonaws.com/"; then
```
Comment (punitmundra) [inline terraform/a2ha-terraform/modules/s3/templates/createS3Bucket.sh:10]: grep -q "http://$1.s3.amazonaws.com/.   is this protocol will be https ?

Comment (ArvinthC3000) [reply terraform/a2ha-terraform/modules/s3/templates/createS3Bucket.sh:10]: Yes.
The following create bucket CLI command `aws s3api create-bucket --bucket $1  --acl private --create-bucket-configuration LocationConstraint=$2` returns the response that has the URL with http protocol

**Review by punitmundra — APPROVED** [review]

File: components/automate-cli/cmd/chef-automate/cleanup.go [hunk 5]
```diff
@@ -133,12 +135,25 @@ func runCleanupCmd(cmd *cobra.Command, args []string) error {
...
-					appendString = appendString + `for i in 1;do i=$PWD;cd /hab/a2_deploy_workspace/terraform/destroy/aws/;cp -r ../../.tf_arch .;cp -r ../../../a2ha.rb ..;terraform apply -var="destroy_bucket=true" -auto-approve;cd $i;done`
+
+				backup_config, err := getTheValueFromA2HARB("backup_config")
+				if err != nil {
+					writer.Error("Error in getting backup_config")
+					return err
+				}
+
+				if backup_config == "s3" && cleanupFlags.force {
+					// bucket_name := config.GetGlobal().GetV1().GetBackups().GetS3().GetBucket().GetName().Value
```
Comment (vivek-yadav) [inline components/automate-cli/cmd/chef-automate/cleanup.go:146]: remove the commented code if not needed.

Comment (ArvinthC3000) [reply components/automate-cli/cmd/chef-automate/cleanup.go]: Removed

File: terraform/a2ha-terraform/modules/s3/main.tf [hunk 6]
```diff
@@ -43,15 +43,22 @@ resource "aws_iam_role_policy_attachment" "AdministratorAccess" {
...
+    command = "chmod 700 ${path.module}/templates/createS3Bucket.sh; ${path.module}/templates/createS3Bucket.sh ${var.aws_s3_bucketName} ${var.aws_region}"
+  }
 }
 
-resource "aws_s3_bucket_acl" "elb_bucket_acl" {
-  bucket = local.log_bucket
-  acl    = "private"
-}
\ No newline at end of file
+# resource "aws_s3_bucket" "createS3bucket" {
```
Comment (vivek-yadav) [inline terraform/a2ha-terraform/modules/s3/main.tf:56]: @ArvinthC3000  please remove this commented code out

**Review by vivekshankar1 — APPROVED** [review]

File: terraform/a2ha-terraform/modules/s3/main.tf (hunk 6 above)
Comment (ArvinthC3000) [reply terraform/a2ha-terraform/modules/s3/main.tf]: Done

**Review by vivek-yadav — APPROVED** [review]

**Pull request merged**